libass/ass_font.c

```diff
@@ -95,14 +95,28 @@ uint32_t ass_font_index_magic(FT_Face face, uint32_t symbol)
     }
 }
 
-static void buggy_font_workaround(FT_Face face)
+static void set_font_metrics(FT_Face face)
 {
-    // Some fonts have zero Ascender/Descender fields in 'hhea' table.
-    // In this case, get the information from 'os2' table or, as
-    // a last resort, from face.bbox.
+    // Mimicking GDI's behavior for asc/desc/height
+    TT_OS2 *os2 = FT_Get_Sfnt_Table(face, ft_sfnt_os2);
+    if (os2 && (os2->usWinAscent + os2->usWinDescent != 0)) {
+        face->ascender  =  os2->usWinAscent;
+        face->descender = -os2->usWinDescent;
+        face->height    = face->ascender - face->descender;
+    }
+
+    // If we didn't have usable Win values in the OS/2 table,
+    // then the values from FreeType will still be in these fields.
+    // It'll use either the OS/2 typo metrics or the hhea ones.
+    // If the font has typo metrics but FreeType didn't use them
+    // (either old FT or USE_TYPO_METRICS not set), we'll try those.
+    // In the case of a very broken font that has none of those options,
+    // we fall back on using face.bbox.
+    // Anything without valid OS/2 Win values isn't supported by VSFilter,
+    // so at this point compatibility's out the window and we're just
+    // trying to render _something_ readable.
     if (face->ascender + face->descender == 0 || face->height == 0) {
```

Comment:
This line is not modified by your commit so it’s by no means your fault, but while we’re at it, I think this is wrong. Shouldn’t this _subtract_ `descender` from `ascender`? As far as I understand surrounding code, `face->ascender` is normally positive and `face->descender` is normally negative, so as it is, this triggers the conditional for fonts whose baseline is exactly in the middle of the glyph, which hardly seems intended.

If you agree, we can probably fix this in a separate commit.

Reply:
Hmmm, looking at 7c901187, I think the intent here was just "ascender == 0 && descender == 0"? Probably makes more sense as `face->ascender - face->descender` though, yeah.

Reply:
Maybe it's better to use intent directly as `face->ascender == 0 && face->descender == 0`?
Also in the similar places above and below it too.

Reply:
Out of scope for the PR, but I think the subtraction is a more sensible thing to do (i.e. "check for 0 space between asc and desc") than checking if both are specifically 0.